src/errors.rs

```diff
@@ -0,0 +1,139 @@
+use std::{
+    cmp::PartialEq,
+    error::Error as StdError,
+    fmt::{self, Display},
+    io,
+};
+
+/// The top-level result type for dealing with Offscale.io stack.
+#[allow(dead_code)]
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// An Error type encapsulating various issues that may come up
+/// in both the expected and unexpected operation of the system.
+#[derive(Debug)]
+pub enum Error {
+    /// Not an error; usually means option None when no error detected.
+    #[allow(dead_code)]
+    NoneError,
+    /// The system has been used in an unsupported way.
+    #[allow(dead_code)]
+    Unsupported(String),
+    /// An unexpected bug has happened. Please open an issue on github!
+    #[allow(dead_code)]
+    ReportableBug(String),
+    /// A read or write error has happened when interacting with the file system.
+    Io(io::Error),
+}
+
+use self::Error::*;
+
+impl Clone for Error {
+    fn clone(&self) -> Error {
+        match self {
+            // Add here variants for new Error enum members when needed.
+            NoneError => NoneError,
+            Unsupported(why) => Unsupported(why.clone()),
+            ReportableBug(what) => ReportableBug(what.clone()),
+            Io(ioe) => Io(std::io::Error::new(ioe.kind(), format!("{:?}", ioe))),
+        }
+    }
+}
+
+impl Eq for Error {}
+
+impl PartialEq for Error {
+    fn eq(&self, other: &Error) -> bool {
+        match *self {
+            // Add here variants for new Error enum members when needed.
+            NoneError => {
+                if let NoneError = *other {
```

Comment:
as per our discussion, make this a `true` rather than a dual branch

Reply:
Changed.